### PR TITLE
Deprecate legacy filter widgets with block alternatives

### LINKED ISSUE #6144: Deprecate old filter widgets in favor of filter blocks
Once we have filter blocks that can interact with the PHP templates, we will be able to deprecate the old filter widgets, that means:

* allowing to insert filter blocks into the widgets screen
* allowing transforms from widgets to the blocks
* hiding the widgets from the inserter

This task should only be done once all other ones in the epic have completed.

### PR DESCRIPTION
**Pull request by tjcafferkey**
This PR gives us a method of soft-deprecating legacy filter widgets for Classic Themes in favour of Filter Blocks which were previously unable to be added to Classic Themes.

Fixes https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/6144

### Screenshots

Screenshot 2022-04-25 at 10 12 23

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests
### Manual Testing

How to test the changes in this Pull Request:

1. Checkout `trunk`.
2. Enable Storefront.
3. Go to Appearance > Widgets and add the following widgets to your sidebar: Filter by Attribute, Filter by Price, Active Filters.
4. Checkout this branch, and run `npm run build`.
5. Go to Appearance Widgets and on your legacy filter widgets, select it to show the toolbar, then select the far left icon.
6. Under the "Transform" dropdown you should be given an option to transform it to the block alternative filter. Do this for all your legacy filters.
7. Now add the Filter by Stock block. There was not a legacy version of this, but it's now available as a block.
8. Check that the legacy widgets you added in step 3 are no longer able to be added.
9. Check on the frontend that all these filters work as expected.

### User Facing Testing
These are steps for user testing (where "user" is someone interacting with this change that is not editing any code).
* [ ] Same as above, or
* [x] See steps below.

1. With an active store using a Classic Theme such as Storefront that has the following legacy widgets in use: Filter by Attribute, Filter by Price, Active Filters.
2. Go to Appearance Widgets and on your legacy filter widgets, select it to show the toolbar, then select the far left icon.
3. Under the "Transform" dropdown you should be given an option to transform it to the block alternative filter. Do this for all your legacy filters.
4. Now add the Filter by Stock block. There was not a legacy version of this, but it's now available as a block.
5. Check that the legacy widgets you added in step 3 are no longer able to be added.
6. Check on the frontend that all these filters work as expected.

### Changelog

> Give merchants the option to convert their legacy filter widgets into filter blocks. Available for Filter by Attribute, Filter by Price, Active Filters and Filter by Stock.

### CODE REVIEW

File: assets/js/blocks/price-filter/index.js [hunk 1]
```diff
@@ -64,10 +64,30 @@ registerBlockType( 'woocommerce/price-filter', {
...
+			{
+				type: 'block',
+				blocks: [ 'core/legacy-widget' ],
+				// We can't transform if raw instance isn't shown in the REST API.
+				isMatch: ( { idBase, instance } ) =>
+					idBase === 'woocommerce_price_filter' && !! instance?.raw,
+				transform: ( { instance } ) =>
+					createBlock( 'woocommerce/price-filter', {
+						showInputFields: true,
+						showFilterButton: false,
```
Comment (Aljullu) [inline assets/js/blocks/price-filter/index.js]: In order to better mimic how the widget works, I would set `showInputFields` to false and `showFilterButton` to true.

File: assets/js/blocks/attribute-filter/index.js [hunk 2]
```diff
@@ -89,6 +89,33 @@ registerBlockType( 'woocommerce/attribute-filter', {
...
+						showCounts: true,
+						queryType: instance?.raw?.query_type || 'or',
+						heading:
+							instance?.raw?.title ||
+							__(
+								'Filter by attribute',
+								'woo-gutenberg-products-block'
+							),
+						headingLevel: 3,
+						displayStyle: instance?.raw?.title || 'list',
```
Comment (Aljullu) [inline assets/js/blocks/attribute-filter/index.js]: Should this be:
```suggestion
						displayStyle: instance?.raw?.display_type || 'list',
```

**Review by Aljullu — COMMENTED** [review]
Overall this is looking great. It's exciting to see that we will be able to deprecate filter widgets! One thing I noticed is that if I add the Filter Products by Rating widget and the Active Filters block, rating filters don't show up. I think that might be worth fixing before deprecating filter widgets. Thoughts @tjcafferkey?

While at some point we would like to create the Filter Products by Rating block, I wonder if we could try and interim solution until then.

**Comment by dinhtungdu** [comment]
> One thing I noticed is that if I add the Filter Products by Rating widget and the Active Filters block, rating filters don't show up. I think that might be worth fixing before deprecating filter widgets. 

I agree and I think this should be addressed in a separate issue/PR. Currently, the Active Filters block supports Price, Stock, and Attribute filters.

File: assets/js/blocks/price-filter/index.js (hunk 1 above)
Comment (tjcafferkey) [reply assets/js/blocks/price-filter/index.js]: Done in https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/6299/commits/af7e7bd74d54f6add17fc461c4bad337ae33a7cc

File: assets/js/blocks/attribute-filter/index.js (hunk 2 above)
Comment (tjcafferkey) [reply assets/js/blocks/attribute-filter/index.js]: Good spot! https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/6299/commits/a79ba94e5c5ed722e6298a638ddd817e7010c78d

**Comment by tjcafferkey** [comment]
>One thing I noticed is that if I add the Filter Products by Rating widget and the Active Filters block, rating filters don't show up. I think that might be worth fixing before deprecating filter widgets.

I would agree with @dinhtungdu here, @Aljullu. I think we should do this in a follow up PR. What do you think?

**Comment by Aljullu** [comment]
> I think we should do this in a follow up PR. What do you think?

I agree with doing it in a separate PR, but IMO we should hold-off in merging this one until that's fixed. Currently the experience is broken if the user uses the Active filters block with the Filter Products by Rating widget, so that could be considered a regression from the previous behavior.

**Comment by tjcafferkey** [comment]
>I agree with doing it in a separate PR, but IMO we should hold-off in merging this one until that's fixed. Currently the experience is broken if the user uses the Active filters block with the Filter Products by Rating widget, so that could be considered a regression from the previous behavior.

Ah yes I see what you're suggesting. OK sounds good I will add a blocked label to this PR until this is fixed.

**Comment by tjcafferkey** [comment]
Blocked until https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/6310 is complete.

**Comment by dinhtungdu** [comment]
FYI https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/6316 which fixes #6310 has been merged.

**Review by dinhtungdu — APPROVED** [review]
LGTM! All filter blocks are working as expected. 🚢

**Comment by dinhtungdu** [comment]
@tjcafferkey E2E test for attribute filter block is passing. The JS unit test is passing as well. Failed E2E tests is related to Checkout block.

**Pull request merged**